File: python/dcbor/src/dcbor/dump.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .string_util import flanked, sanitized
from .varint import MajorType, encode_varint

if TYPE_CHECKING:
    from .cbor import CBOR
    from .tag import Tag
    from .tags_store import TagsStore

_NO_TAGS = object()


class _DumpItem:
    __slots__ = ("level", "data", "note")

    def __init__(
        self,
        level: int,
        data: list[bytes],
        note: str | None,
    ) -> None:
        self.level = level
        self.data = data
        self.note = note

    def format_first_column(self) -> str:
        indent = " " * (self.level * 4)
        hex_parts = [d.hex() for d in self.data if len(d) > 0]
        hex_str = " ".join(hex_parts)
        return indent + hex_str

    def format(self, note_column: int) -> str:
        col1 = self.format_first_column()
        if self.note is not None:
            padding_count = max(1, min(39, note_column) - len(col1) + 1)
            padding = " " * padding_count
            col2 = f"# {self.note}"
            return col1 + padding + col2
        return col1


def _dump_items(cbor: CBOR, level: int, tags_store: TagsStore | None) -> list[_DumpItem]:
    from .cbor import CBORCase

    match cbor.case:
        case CBORCase.UNSIGNED:
            return [_DumpItem(
                level,
                [cbor.to_cbor_data()],
                f"unsigned({cbor.value})",
            )]

        case CBORCase.NEGATIVE:
            return [_DumpItem(
                level,
                [cbor.to_cbor_data()],
                f"negative({-1 - cbor.value})",
            )]

        case CBORCase.BYTE_STRING:
            bs = cbor.value
            data = bs.data
            header = encode_varint(len(data), MajorType.BYTE_STRING)
            items = [_DumpItem(level, [header], f"bytes({len(data)})")]
            if len(data) > 0:
                note: str | None = None
                try:
                    text = data.decode("utf-8")
                    san = sanitized(text)
                    if san is not None:
                        note = flanked(san, '"', '"')
                except (UnicodeDecodeError, ValueError):
                    pass
                items.append(_DumpItem(level + 1, [data], note))
            return items

        case CBORCase.TEXT:
            text = cbor.value
            utf8_data = text.encode("utf-8")
            header = encode_varint(len(utf8_data), MajorType.TEXT)
            header_data = [bytes([header[0]]), header[1:]]
            return [
                _DumpItem(level, header_data, f"text({len(utf8_data)})"),
                _DumpItem(level + 1, [utf8_data], flanked(text, '"', '"')),
            ]

        case CBORCase.SIMPLE:
            data = cbor.value.cbor_data()
            note = str(cbor.value)
            return [_DumpItem(level, [data], note)]

        case CBORCase.TAGGED:
            tag, item = cbor.value
            header = encode_varint(tag.value, MajorType.TAGGED)
            header_data = [bytes([header[0]]), header[1:]]
            note_parts = [f"tag({tag.value})"]
            assigned = _get_assigned_name_dump(tag, tags_store)
            if assigned is not None:
                note_parts.append(assigned)
            tag_note = " ".join(note_parts)
            result = [_DumpItem(level, header_data, tag_note)]
            result.extend(_dump_items(item, level + 1, tags_store))
            return result

        case CBORCase.ARRAY:
            arr = cbor.value
            header = encode_varint(len(arr), MajorType.ARRAY)
            header_data = [bytes([header[0]]), header[1:]]
            result = [_DumpItem(level, header_data, f"array({len(arr)})")]
            for child in arr:
                result.extend(_dump_items(child, level + 1, tags_store))
            return result

        case CBORCase.MAP:
            m = cbor.value
            entry_count = len(m)
            header = encode_varint(entry_count, MajorType.MAP)
            header_data = [bytes([header[0]]), header[1:]]
            result = [_DumpItem(level, header_data, f"map({entry_count})")]
            for key, val in m.iter():
                result.extend(_dump_items(key, level + 1, tags_store))
                result.extend(_dump_items(val, level + 1, tags_store))
            return result

    return []


def _get_assigned_name_dump(tag: Tag, tags_store: TagsStore | None) -> str | None:
    if tags_store is None:
        from .tags_store import _get_global_tags
        store = _get_global_tags()
        return store.assigned_name_for_tag(tag)
    if tags_store is _NO_TAGS:
        return None
    return tags_store.assigned_name_for_tag(tag)


def hex_annotated_impl(cbor: CBOR, *, tags_store: TagsStore | None = None) -> str:
    items = _dump_items(cbor, 0, tags_store)
    if not items:
        return ""
    note_column = max(len(item.format_first_column()) for item in items)
    note_column = ((note_column + 4) & ~3) - 1
    lines = [item.format(note_column) for item in items]
    return "\n".join(lines)

File: csharp/DCbor/DCbor/Dump.cs
namespace BlockchainCommons.DCbor;

/// <summary>
/// Options controlling hex dump formatting.
/// </summary>
public class HexFormatOptions
{
    public bool Annotate { get; set; }
    public TagsStoreOption Tags { get; set; } = TagsStoreOption.DefaultGlobal;

    public HexFormatOptions WithAnnotate(bool v) { Annotate = v; return this; }
    public HexFormatOptions WithTags(TagsStoreOption v) { Tags = v; return this; }
}

/// <summary>
/// Hex-annotated dump rendering for CBOR values.
/// </summary>
public static class CborDump
{
    /// <summary>Returns the hex-encoded CBOR bytes.</summary>
    public static string HexDump(this Cbor cbor)
    {
        return cbor.Hex();
    }

    /// <summary>Returns annotated hex dump.</summary>
    public static string HexAnnotated(this Cbor cbor)
    {
        return cbor.HexOpt(new HexFormatOptions { Annotate = true });
    }

    /// <summary>Returns hex dump with custom options.</summary>
    public static string HexOpt(this Cbor cbor, HexFormatOptions opts)
    {
        if (!opts.Annotate)
            return cbor.Hex();

        var items = DumpItems(cbor, 0, opts);
        int noteColumn = 0;
        foreach (var item in items)
        {
            int len = item.FormatFirstColumn().Length;
            if (len > noteColumn) noteColumn = len;
        }
        // Round up to nearest multiple of 4
        noteColumn = ((noteColumn + 4) & ~3) - 1;

        var lines = items.Select(x => x.Format(noteColumn)).ToList();
        return string.Join("\n", lines);
    }

    private static List<DumpItem> DumpItems(Cbor cbor, int level, HexFormatOptions opts)
    {
        switch (cbor.Case)
        {
            case CborCase.UnsignedCase u:
                return new List<DumpItem>
                {
                    new(level, new List<byte[]> { cbor.ToCborData() },
                        $"unsigned({u.Value})")
                };

            case CborCase.NegativeCase n:
                return new List<DumpItem>
                {
                    new(level, new List<byte[]> { cbor.ToCborData() },
                        $"negative({-1 - (Int128)n.Value})")
                };

            case CborCase.ByteStringCase bs:
            {
                var header = Varint.EncodeVarInt((ulong)bs.Value.Length, MajorType.ByteString);
                var items = new List<DumpItem>
                {
                    new(level, new List<byte[]> { header },
                        $"bytes({bs.Value.Length})")
                };
                if (!bs.Value.IsEmpty)
                {
                    string? note = null;
                    try
                    {
                        var strictUtf8 = new System.Text.UTF8Encoding(false, true);
                        string str = strictUtf8.GetString(bs.Value.DataRef);
                        var sanitized = StringUtil.Sanitized(str);
                        if (sanitized != null)
                            note = StringUtil.Flanked(sanitized, "\"", "\"");
                    }
                    catch { /* not valid UTF-8 */ }
                    items.Add(new DumpItem(level + 1, new List<byte[]> { bs.Value.ToArray() }, note));
                }
                return items;
            }

            case CborCase.TextCase t:
            {
                var header = Varint.EncodeVarInt((ulong)System.Text.Encoding.UTF8.GetByteCount(t.Value), MajorType.Text);
                var utf8 = System.Text.Encoding.UTF8.GetBytes(t.Value);
                var headerData = new List<byte[]>();
                headerData.Add(new[] { header[0] });
                if (header.Length > 1)
                    headerData.Add(header[1..]);
                return new List<DumpItem>
                {
                    new(level, headerData, $"text({utf8.Length})"),
                    new(level + 1, new List<byte[]> { utf8 },
                        StringUtil.Flanked(t.Value, "\"", "\""))
                };
            }

            case CborCase.SimpleCase s:
            {
                var data = s.Value.CborData();
                return new List<DumpItem>
                {
                    new(level, new List<byte[]> { data }, s.Value.ToString())
                };
            }

            case CborCase.TaggedCase tg:
            {
                var header = Varint.EncodeVarInt(tg.Tag.Value, MajorType.Tagged);
                var headerData = new List<byte[]> { new[] { header[0] } };
                if (header.Length > 1)
                    headerData.Add(header[1..]);

                var noteComponents = new List<string> { $"tag({tg.Tag.Value})" };
                switch (opts.Tags)
                {
                    case TagsStoreOption.Global:
                    {
                        var name = GlobalTags.WithTags(store => store.AssignedNameForTag(tg.Tag));
                        if (name != null) noteComponents.Add(name);
                        break;
                    }
                    case TagsStoreOption.Custom custom:
                    {
                        var name = custom.Store.AssignedNameForTag(tg.Tag);
                        if (name != null) noteComponents.Add(name);
                        break;
                    }
                }

                var items = new List<DumpItem>
                {
                    new(level, headerData, string.Join(" ", noteComponents))
                };
                items.AddRange(DumpItems(tg.Item, level + 1, opts));
                return items;
            }

            case CborCase.ArrayCase a:
            {
                var header = Varint.EncodeVarInt((ulong)a.Value.Count, MajorType.Array);
                var headerData = new List<byte[]> { new[] { header[0] } };
                if (header.Length > 1)
                    headerData.Add(header[1..]);

                var items = new List<DumpItem>
                {
                    new(level, headerData, $"array({a.Value.Count})")
                };
                foreach (var item in a.Value)
                    items.AddRange(DumpItems(item, level + 1, opts));
                return items;
            }

            case CborCase.MapCase m:
            {
                var header = Varint.EncodeVarInt((ulong)m.Value.Count, MajorType.Map);
                var headerData = new List<byte[]> { new[] { header[0] } };
                if (header.Length > 1)
                    headerData.Add(header[1..]);

                var items = new List<DumpItem>
                {
                    new(level, headerData, $"map({m.Value.Count})")
                };
                foreach (var (key, value) in m.Value)
                {
                    items.AddRange(DumpItems(key, level + 1, opts));
                    items.AddRange(DumpItems(value, level + 1, opts));
                }
                return items;
            }

            default:
                throw new InvalidOperationException();
        }
    }

    private sealed class DumpItem
    {
        private readonly int _level;
        private readonly List<byte[]> _data;
        private readonly string? _note;

        internal DumpItem(int level, List<byte[]> data, string? note)
        {
            _level = level;
            _data = data;
            _note = note;
        }

        internal string Format(int noteColumn)
        {
            string col1 = FormatFirstColumn();
            if (_note == null)
                return col1;

            long paddingCount = Math.Max(1, Math.Min(39, noteColumn) - col1.Length + 1);
            string padding = new string(' ', (int)paddingCount);
            return $"{col1}{padding}# {_note}";
        }

        internal string FormatFirstColumn()
        {
            string indent = new string(' ', _level * 4);
            var hexParts = _data
                .Select(d => Convert.ToHexString(d).ToLowerInvariant())
                .Where(h => h.Length > 0)
                .ToList();
            return indent + string.Join(" ", hexParts);
        }
    }
}
